File: sbe/binance/spot_3_2/generated-php/spot_sbe/AccountResponse.php
<?php
/**
 * Generated SBE (Simple Binary Encoding) message codec.
 */

class Balances
{
    public int|float|null $exponent = null;
    public int|float|null $free = null;
    public int|float|null $locked = null;
}

class Permissions
{
}

class ReduceOnlyAssets
{
}

class AccountResponse
{
    public const TEMPLATE_ID = 400;
    public const SCHEMA_ID = 3;
    public const SCHEMA_VERSION = 2;
    public const BLOCK_LENGTH = 64;

    public int|float|array|null $commissionExponent = null;
    public int|float|array|null $commissionRateMaker = null;
    public int|float|array|null $commissionRateTaker = null;
    public int|float|array|null $commissionRateBuyer = null;
    public int|float|array|null $commissionRateSeller = null;
    public int|float|array|null $canTrade = null;
    public int|float|array|null $canWithdraw = null;
    public int|float|array|null $canDeposit = null;
    public int|float|array|null $brokered = null;
    public int|float|array|null $requireSelfTradePrevention = null;
    public int|float|array|null $preventSor = null;
    public int|float|array|null $updateTime = null;
    public int|float|array|null $accountType = null;
    public int|float|array|null $tradeGroupId = null;
    public int|float|array|null $uid = null;
    public array $balances = [];
    public array $permissions = [];
    public array $reduceOnlyAssets = [];

    private function decodeBalancesGroup(string $data, int &$offset): array
    {
        $blockLength = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;
        $numInGroup = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;

        $items = [];
        for ($i = 0; $i < $numInGroup; $i++) {
            $itemStart = $offset;
            $item = new Balances();

            $item->exponent = unpack('c', substr($data, $offset, 1))[1];
            $offset += 1;
            $item->free = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->locked = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;

            // Skip to next block for forward compatibility
            $offset = $itemStart + $blockLength;
            $items[] = $item;
        }

        return $items;
    }

    private function decodePermissionsGroup(string $data, int &$offset): array
    {
        $blockLength = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;
        $numInGroup = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;

        $items = [];
        for ($i = 0; $i < $numInGroup; $i++) {
            $itemStart = $offset;
            $item = new Permissions();


            // Skip to next block for forward compatibility
            $offset = $itemStart + $blockLength;
            $items[] = $item;
        }

        return $items;
    }

    private function decodeReduceOnlyAssetsGroup(string $data, int &$offset): array
    {
        $blockLength = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;
        $numInGroup = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;

        $items = [];
        for ($i = 0; $i < $numInGroup; $i++) {
            $itemStart = $offset;
            $item = new ReduceOnlyAssets();


            // Skip to next block for forward compatibility
            $offset = $itemStart + $blockLength;
            $items[] = $item;
        }

        return $items;
    }

    public function encode(): string
    {
        $buffer = '';

        if ($this->commissionExponent !== null) {
            $buffer .= pack('c', $this->commissionExponent);
        }
        if ($this->commissionRateMaker !== null) {
            $buffer .= pack('q', $this->commissionRateMaker);
        }
        if ($this->commissionRateTaker !== null) {
            $buffer .= pack('q', $this->commissionRateTaker);
        }
        if ($this->commissionRateBuyer !== null) {
            $buffer .= pack('q', $this->commissionRateBuyer);
        }
        if ($this->commissionRateSeller !== null) {
            $buffer .= pack('q', $this->commissionRateSeller);
        }
        if ($this->updateTime !== null) {
            $buffer .= pack('q', $this->updateTime);
        }
        if ($this->tradeGroupId !== null) {
            $buffer .= pack('q', $this->tradeGroupId);
        }
        if ($this->uid !== null) {
            $buffer .= pack('q', $this->uid);
        }

        return $buffer;
    }

    public function decode(string $data): void
    {
        $offset = 0;

        $this->commissionExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->commissionRateMaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->commissionRateTaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->commissionRateBuyer = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->commissionRateSeller = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->updateTime = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->tradeGroupId = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->uid = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;

        // Skip to end of block for forward compatibility
        $offset = 64;

        $this->balances = $this->decodeBalancesGroup($data, $offset);
        $this->permissions = $this->decodePermissionsGroup($data, $offset);
        $this->reduceOnlyAssets = $this->decodeReduceOnlyAssetsGroup($data, $offset);
    }
}

File: sbe/binance/spot_3_2/generated-python/spot_sbe/AccountResponse.py
"""Generated SBE (Simple Binary Encoding) message codec."""

import struct
from typing import List, Optional, Tuple, Set
from io import BytesIO

class Balances:
    """Repeating group item."""

    def __init__(self):
        self.exponent = None
        self.free = None
        self.locked = None

class Permissions:
    """Repeating group item."""

    def __init__(self):

class ReduceOnlyAssets:
    """Repeating group item."""

    def __init__(self):

class AccountResponse:
    """SBE message: AccountResponse."""

    TEMPLATE_ID = 400
    SCHEMA_ID = 3
    SCHEMA_VERSION = 2
    BLOCK_LENGTH = 64

    def __init__(self):
        self.commission_exponent = None
        self.commission_rate_maker = None
        self.commission_rate_taker = None
        self.commission_rate_buyer = None
        self.commission_rate_seller = None
        self.can_trade = None
        self.can_withdraw = None
        self.can_deposit = None
        self.brokered = None
        self.require_self_trade_prevention = None
        self.prevent_sor = None
        self.update_time = None
        self.account_type = None
        self.trade_group_id = None
        self.uid = None
        self.balances = []
        self.permissions = []
        self.reduce_only_assets = []

    def _decode_balances_group(self, data: bytes, offset: int) -> Tuple[List[Balances], int]:
        """Decode repeating group."""
        pos = offset

        block_length = struct.unpack_from('<H', data, pos)[0]
        pos += 2
        num_in_group = struct.unpack_from('<I', data, pos)[0]
        pos += 4

        items = []
        for _ in range(num_in_group):
            item_start = pos
            item = Balances()

            item.exponent = struct.unpack_from('<b', data, pos)[0]
            pos += 1
            item.free = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.locked = struct.unpack_from('<q', data, pos)[0]
            pos += 8

            # Skip to next block for forward compatibility
            pos = item_start + block_length
            items.append(item)

        return (items, pos)

    def _decode_permissions_group(self, data: bytes, offset: int) -> Tuple[List[Permissions], int]:
        """Decode repeating group."""
        pos = offset

        block_length = struct.unpack_from('<H', data, pos)[0]
        pos += 2
        num_in_group = struct.unpack_from('<I', data, pos)[0]
        pos += 4

        items = []
        for _ in range(num_in_group):
            item_start = pos
            item = Permissions()


            # Skip to next block for forward compatibility
            pos = item_start + block_length
            items.append(item)

        return (items, pos)

    def _decode_reduce_only_assets_group(self, data: bytes, offset: int) -> Tuple[List[ReduceOnlyAssets], int]:
        """Decode repeating group."""
        pos = offset

        block_length = struct.unpack_from('<H', data, pos)[0]
        pos += 2
        num_in_group = struct.unpack_from('<I', data, pos)[0]
        pos += 4

        items = []
        for _ in range(num_in_group):
            item_start = pos
            item = ReduceOnlyAssets()


            # Skip to next block for forward compatibility
            pos = item_start + block_length
            items.append(item)

        return (items, pos)

    def encode(self) -> bytes:
        """Encode the message to bytes."""
        buffer = BytesIO()

        if self.commission_exponent is not None:
            buffer.write(struct.pack('<b', self.commission_exponent))
        if self.commission_rate_maker is not None:
            buffer.write(struct.pack('<q', self.commission_rate_maker))
        if self.commission_rate_taker is not None:
            buffer.write(struct.pack('<q', self.commission_rate_taker))
        if self.commission_rate_buyer is not None:
            buffer.write(struct.pack('<q', self.commission_rate_buyer))
        if self.commission_rate_seller is not None:
            buffer.write(struct.pack('<q', self.commission_rate_seller))
        if self.update_time is not None:
            buffer.write(struct.pack('<q', self.update_time))
        if self.trade_group_id is not None:
            buffer.write(struct.pack('<q', self.trade_group_id))
        if self.uid is not None:
            buffer.write(struct.pack('<q', self.uid))

        return buffer.getvalue()

    def decode(self, data: bytes) -> None:
        """Decode the message from bytes."""
        pos = 0

        self.commission_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1
        self.commission_rate_maker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.commission_rate_taker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.commission_rate_buyer = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.commission_rate_seller = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.update_time = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.trade_group_id = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.uid = struct.unpack_from('<q', data, pos)[0]
        pos += 8

        # Skip to end of block for forward compatibility
        pos = 64

        self.balances, pos = self._decode_balances_group(data, pos)
        self.permissions, pos = self._decode_permissions_group(data, pos)
        self.reduce_only_assets, pos = self._decode_reduce_only_assets_group(data, pos)
